### ISSUE TITLE
[FastestDet] The absolute error of `AveragePool`, `GlobalAveragePool` is quite large

### ISSUE THREAD

**Issue by PINTO0309** [PINTO0309/onnx2tf#124]
### Issue Type

Others

### onnx2tf version number

1.5.12

### onnx version number

1.13.0

### tensorflow version number

2.10.0

### Download URL for ONNX

https://github.com/PINTO0309/onnx2tf/releases/download/1.1.28/fastestdet.onnx
https://github.com/PINTO0309/onnx2tf/releases/download/1.1.28/ppmattingv2_stdc1_human_480x640.onnx

### Parameter Replacement JSON

```json
N/A
```

### Description

1. Personal
2. The absolute error of `AveragePool` is quite large. In addition, there is almost no error in other OPs.
    ```
    onnx2tf -i fastestdet.onnx -cotof -cotoa 1e-4
    ```
    ![image](https://user-images.githubusercontent.com/33194443/212523271-ef5ccf5d-1ec7-433b-8d72-e5de32e54e4e.png)
    ![image](https://user-images.githubusercontent.com/33194443/212523247-6f18226b-c006-49df-b01c-01f9144eb1ec.png)
4. At this time I do not know the cause of the larger error.
5. I want to generate a tfite model with small accuracy degradation.

**Comment by PINTO0309** [comment]
Oddly enough, however, the `GlobalAveragePool` in regnet_x_400mf matched the final output. The only difference I can see is whether the input is square or rectangular.
https://github.com/PINTO0309/onnx2tf/releases/download/1.1.28/regnet_x_400mf.onnx
![image](https://user-images.githubusercontent.com/33194443/212547071-895ea1d7-1684-4aae-9fca-30550f0bac35.png)

![image](https://user-images.githubusercontent.com/33194443/212547045-b11c3917-7ef4-4749-b3f5-961924db5815.png)

**Comment by PINTO0309** [comment]
For ONNX, when `count_include_pad : int (default is 0) =0` and `padding size >0`, zero-padded edge portions are excluded from the average value calculation. Therefore, if zero padding is applied to the edges first and processing is performed using TensorFlow's `AveragePooling2D`, the average value is calculated with a total of 9 pixel values including the zero-padded pixel, and the values are considered to be inconsistent.

https://github.com/microsoft/onnxruntime/blob/0c6037b5abe571fc43a55ef7a9d2f846820fbe5d/onnxruntime/test/mlas/unittest/test_pool2d.h#L148-L210

- `kernel_shape = [3,3]`, `padding = [1,1,1,1]`, `strides = [2,2]`
  ![image](https://user-images.githubusercontent.com/33194443/212578264-9bf6e1a0-c49c-43f3-a867-aa6538bf94eb.png)

**Comment by PINTO0309** [comment]
This is an experimental implementation of `AveragePool`, forcing it to meet ONNX specifications. The accuracy is guaranteed, but the implementation is very redundant. I understand the critical specification differences between ONNX and TensorFlow, but trying to force the specifications to match generates an incredibly redundant model.

No merging until there is a realistic implementation proposal.

commit: https://github.com/PINTO0309/onnx2tf/commit/cab9359d28e73aa4d1bfdf6c9b5c59d6fd3adbfc

![image](https://user-images.githubusercontent.com/33194443/212695683-2af127b7-790a-4eb8-b813-25fe47d254be.png)

**Comment by Hyunseok-Kim0** [comment]
![Screenshot from 2023-01-17 13-23-21](https://user-images.githubusercontent.com/34959032/212814634-c68bcc43-aa5d-4540-98f7-fbaec9152152.png)

<img src="https://user-images.githubusercontent.com/34959032/212814866-fdc11379-fe80-4280-8191-e68f1db8942e.png" height="400">

The images above are outputs of average pooling 2d from tflite, onnx, and tflite output divided by onnx output. As shown in image, output is same except the padded edges. I think we can get same output by padding slightly more and slice useless part.
By the way, what was the bug you mentioned in #121?

**Comment by PINTO0309** [comment]
>By the way, what was the bug you mentioned in https://github.com/PINTO0309/onnx2tf/pull/121?

This is exactly the process described in this issue. I was trying to reduce the error wildly. Padding with edge values rather than filling with zeros just tended to reduce the error. 

I have not been very constructive in that pull request. :face_in_clouds:

**Comment by Hyunseok-Kim0** [comment]
So is that mean Mirrored pad is not necessary if this problem can be solved using extra pad and slice?

**Comment by PINTO0309** [comment]
That is correct.

**Comment by Hyunseok-Kim0** [comment]
<img src="https://user-images.githubusercontent.com/34959032/213062860-98899a7d-a973-4e79-b29f-358b94ee1fb1.png" width="200">

I modified `AveragePool.py` and converted tflite has structure shown above. The value first row or column of each axis is multiplied by specific number, which is `kernel_shape / strides`. If last row or column is not dropped during pooling, that one is also multiplied. The shape is quite nasty, but the output looks same.

However, there is two problem.
1. It looks like `count_include_pad` is always False.
When pytorch model is exported to onnx, `count_include_pad` is always `False`. The option itself always doesn't exist in attributes list of graph node. I am not sure whether this is normal or not.

2. FastestDet tflite and onnx has different output.
![Screenshot from 2023-01-18 11-14-23](https://user-images.githubusercontent.com/34959032/213065131-a7ab0194-3725-46f8-8d49-d359806460d2.png)
I think this is not related to average pooling. There is difference in output of `Transpose_263`, but not detected in `onnx_tf_tensor_validation`. When I used https://www.traveloffpath.com/wp-content/uploads/2021/03/Crowd-of-people-some-in-masks-and-some-not.jpg for testing input, `rearrange(tflite_output, "n h w c -> n c h w")[:, -23:-22, ...], onnx_output[:, -23:-22, ...]` shows result above.

I am still testing the validity. Do you have any other opinion about the implementation? I tried to find out if there is more elegant way to reduce the extra part after pooling layer, but didn't work.

**Comment by PINTO0309** [comment]
Thanks for the great idea. :+1: 

> Come to think of it, I didn't test it, but if that's true, it sounds like a bug in PyTorch.

https://pytorch.org/docs/stable/generated/torch.nn.AvgPool2d.html

> There is difference in output of Transpose_263, but not detected in onnx_tf_tensor_validation.

`Conv2D` may cause a small error. But the errors you've shown appear to be much less than the overall errors that were occurring before. While it is obviously desirable to have almost zero errors, I think there are similar problems with layers other than `AveragePool2D`.

> Do you have any other opinion about the implementation?

I can't think of a better idea than yours at the moment. As far as I know, we are the only ones seriously addressing this issue.

**Comment by Hyunseok-Kim0** [comment]
https://github.com/PINTO0309/onnx2tf/blob/27d635f5a962b14f9655aa48e5c5660083217cc2/onnx2tf/ops/AveragePool.py#L129-L134
It looks line 133 is not appropriate for pooling operation. When stride is larger than 1, this line is always True. After that line, padding mode is always 'valid'.

Also, according to https://stackoverflow.com/questions/50543862/tensorflow-tf-nn-avg-pool-with-same-padding-does-not-average-over-padded-pi, tensorflow default action looks like `count_include_pad=False`. I guess, In this case, explicit padding and pad mode 'valid' caused output difference between onnx and tflite.

Keep trying to fix, but this problem is quite complicated than I thought.

**Comment by PINTO0309** [comment]
`MaxPool` and `AveragePool` describe rather crude implementations to normalize the output of YOLOX and YOLOvNN. My initial implementation was taken from onnx-tensorflow, but it seemed to lack consideration of padding. 

The current implementation has a lot of unnecessary and redundant processes written in it, and it would be better to remove unnecessary processes as they become available. I spent a lot of time on Pooling operations alone and could not come up with a good implementation.

**Comment by PINTO0309** [comment]
[BUG] onnxsim simplifies pad + average pool into a single average pool without handling the pad border mode correctly

https://github.com/daquexian/onnx-simplifier/issues/244

**Comment by PINTO0309** [comment]
There appears to be a problem with the padding composition algorithm in onnx-simplifier. As a test, I generated an ONNX file for a FastestDet that did not implement onnxsim. 
[FastestDet_no_simplify.onnx.zip](https://github.com/PINTO0309/onnx2tf/files/10452203/FastestDet_no_simplify.onnx.zip)
![image](https://user-images.githubusercontent.com/33194443/213325466-0bb778ee-d1b9-40d3-bf41-823a090c9a6f.png)

**Comment by PINTO0309** [comment]
Thanks to Hyunseok-Kim0 for the help in eliminating the error. Thank you very much.
https://github.com/PINTO0309/onnx2tf/releases/tag/1.5.21

**Comment by daquexian** [comment]
FYI, https://github.com/daquexian/onnx-simplifier/issues/244 has been resolved in the latest version

**Comment by PINTO0309** [comment]
@daquexian Thank you!
